File: go-backend-interview-prep/01_go_basics/07_for_loops/main.py
# Python equivalent of For Loops (compare with main.go)

def main():
    # --- C-style for -> range ---
    print("--- C-style (range) ---")
    for i in range(5):
        print(i, end=" ")
    print()

    # --- While-style ---
    print("--- While-style ---")
    n = 1
    while n < 50:
        n *= 2
    print("n doubled to:", n)

    # --- Infinite loop with break ---
    print("--- Infinite + break ---")
    counter = 0
    while True:
        if counter >= 3:
            break
        print("counter:", counter)
        counter += 1

    # --- Continue (skip iteration) ---
    print("--- Continue (skip even) ---")
    for i in range(6):
        if i % 2 == 0:
            continue
        print(i, end=" ")
    print()

    # --- Iterate over a string (by character, not byte) ---
    print("--- Iterate string ---")
    for idx, ch in enumerate("Go!"):
        print(f"index={idx} char={ch}")

    # --- Ignoring index with _ ---
    print("--- Ignore index ---")
    for _, ch in enumerate("abc"):
        print(ch, end=" ")
    print()
    # Or simply: for ch in "abc":

    # --- Nested loop with break-out ---
    # Python has no labeled break. Use a flag or function.
    print("--- Break from nested ---")
    done = False
    for i in range(3):
        for j in range(3):
            if i == 1 and j == 1:
                done = True
                break
            print(f"({i},{j})", end=" ")
        if done:
            break
    print()


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/01_go_basics/07_for_loops/main.go
package main

import "fmt"

func main() {
	// --- C-style for loop ---
	fmt.Println("--- C-style ---")
	for i := 0; i < 5; i++ {
		fmt.Print(i, " ")
	}
	fmt.Println()

	// --- While-style (condition only) ---
	fmt.Println("--- While-style ---")
	n := 1
	for n < 50 {
		n *= 2
	}
	fmt.Println("n doubled to:", n)

	// --- Infinite loop with break ---
	fmt.Println("--- Infinite + break ---")
	counter := 0
	for {
		if counter >= 3 {
			break
		}
		fmt.Println("counter:", counter)
		counter++
	}

	// --- Continue (skip iteration) ---
	fmt.Println("--- Continue (skip even) ---")
	for i := 0; i < 6; i++ {
		if i%2 == 0 {
			continue
		}
		fmt.Print(i, " ")
	}
	fmt.Println()

	// --- Range over a string ---
	// Returns byte index and rune (Unicode code point)
	fmt.Println("--- Range over string ---")
	for idx, ch := range "Go!" {
		fmt.Printf("index=%d char=%c\n", idx, ch)
	}

	// --- Ignoring index with _ ---
	fmt.Println("--- Ignore index ---")
	for _, ch := range "abc" {
		fmt.Printf("%c ", ch)
	}
	fmt.Println()

	// --- Nested loop with labeled break ---
	fmt.Println("--- Labeled break ---")
outer:
	for i := 0; i < 3; i++ {
		for j := 0; j < 3; j++ {
			if i == 1 && j == 1 {
				break outer // breaks out of BOTH loops
			}
			fmt.Printf("(%d,%d) ", i, j)
		}
	}
	fmt.Println()
}
